manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,451 @@
+	// published port to dynamically assigning a published port. In that case,
+	// the actual state of the object will have a published port, but the
+	// user's spec wants a new (dynamically allocated) port, just like if we
+	// had dynamically assigned the port.
+	//
+	// Luckily for use, we keep a copy of the spec on the Endpoint object,
+	// which we own, which means we have the old endpoint spec around and can
+	// compare the user's specs. Then, all we have to do is see if the
+	// published port changed.
+
+	// So, basically, here's what we're going to do: we're going to create a
+	// new list of PortConfigs. This will be the final list that gets put into
+	// the object endpoint object
+	finalPorts := make([]*api.PortConfig, len(spec.Ports))
+
+	// first, we need to "recover" any dynamically assigned publish ports. to
+	// do this:
+	//   1. go through all of the ports in the new spec.
+	//   2. if a port doesn't have a publish port assigned
+	//     a. go through all of the ports in the old, checking if one matches
+	//     b. if so, that port has a dynamic port assignment
+	//        i. go through every port in the old object's ports. if every
+	//           component of the port is the same EXCEPT the PublishedPort,
+	//           copy that published port into the port assignments.
+	for i, spec := range spec.Ports {
+		// check if the published port or target port is off the end of the
+		// allowed port range
+		if spec.PublishedPort > masterPortEnd {
+			return nil, ErrInvalidEndpoint{fmt.Sprintf("published port %v isn't in the valid port range", spec.PublishedPort)}
+		}
+		if spec.TargetPort > masterPortEnd {
+			return nil, ErrInvalidEndpoint{fmt.Sprintf("target port %v isn't in the valid port range", spec.TargetPort)}
+		}
+		// copy the port from the spec into the final ports list
+		finalPorts[i] = spec.Copy()
+		// if the publish mode is host, we're done
+		if spec.PublishMode == api.PublishModeHost {
+			continue
+		}
+		// if the published port of the spec is given by the user, we're done
+		if spec.PublishedPort != 0 {
+			continue
+		}
+		// check if there is an old spec. If not, we don't have to go looking
+		// for a previously assigned dynamic port
+		if endpoint.Spec == nil {
+			continue
+		}
+		// go through all of the ports in the old spec, checking if one matches
+		for _, old := range endpoint.Spec.Ports {
+			if portsEqual(old, spec) {
+				// if the do match, find the port from the object
+				for _, p := range endpoint.Ports {
+					if PortsMostlyEqual(spec, p) {
+						// and take it's PublishedPort assignment for the final
+						// port assignment
+						finalPorts[i].PublishedPort = p.PublishedPort
+					}
+				}
+			}
+		}
+	}
+
+	// to avoid altering the "real" state of the ports map, we use the proposal
+	// object, which contains the proposed changes to the Allocator. this
+	// means if there is a failure in the caller after calling Allocate, the
+	// caller can discard the changes
+	prop := &proposal{
```

Comment:
Non-blocking:  I was wondering if we could just call Deallocate here, and modify the proposal that it returned with an `allocate` map?  Since otherwise the code is the same.